### PR TITLE
Improved HTTP response length handling

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3907

There's a few error handling codepaths that aren't handled, like "mutable changed out from under you" or "there's a bug!". Can try to figure something out if reviewer feels it's useful.

### CODE REVIEW

**Comment by itamarst** [comment]
BOOOOOO

```
2022-07-20T18:45:16.4509660Z coveralls.exception.CoverallsException: Could not submit coverage: 504 Server Error: Gateway Time-out for url: https://coveralls.io/api/v1/jobs
```

File: src/allmydata/storage/http_client.py [hunk 1]
```diff
@@ -461,13 +463,40 @@ def read_share_chunk(
...
+            raise ValueError(
+                f"Length of response sent from server ({actual_length}) "
+                + f"didn't match Content-Range header ({supposed_length})"
+            )
+        body.seek(0)
+        return body.read()
     else:
+        # Technically HTTP allows sending an OK with full body under these
+        # circumstances, but the server is not designed to do that so we ignore
+        # than possibility for now...
```
Comment (exarkun) [inline src/allmydata/storage/http_client.py:499]:
```suggestion
        # that possibility for now...
```

File: src/allmydata/storage/http_client.py [hunk 2]
```diff
@@ -461,13 +463,40 @@ def read_share_chunk(
...
+        if supposed_length > length:
+            raise ValueError("Server sent more than we asked for?!")
+        # It might also send less than we asked for. That's (probably) OK, e.g.
+        # if we went past the end of the file.
+        body = yield limited_content(response, supposed_length)
+        body.seek(0, SEEK_END)
+        actual_length = body.tell()
+        if actual_length != supposed_length:
+            # Most likely a mutable that got changed out from under us, but
+            # concievably could be a bug...
```
Comment (exarkun) [inline src/allmydata/storage/http_client.py:489]:
```suggestion
            # conceivably could be a bug...
```

File: src/allmydata/storage/immutable.py [hunk 3]
```diff
@@ -199,8 +199,13 @@ def __init__(
                 raise UnknownImmutableContainerVersionError(filename, version)
             self._num_leases = num_leases
             self._lease_offset = filesize - (num_leases * self.LEASE_SIZE)
+            self._length = filesize - 0xc - (num_leases * self.LEASE_SIZE)
+
         self._data_offset = 0xc
 
+    def get_length(self):
+        return self._length
```
Comment (exarkun) [inline src/allmydata/storage/immutable.py:207]: a docstring here would be nice

File: src/allmydata/storage/immutable.py [hunk 4]
```diff
@@ -544,6 +549,9 @@ def advise_corrupt_share(self, reason):
                                             self.shnum,
                                             reason)
 
+    def get_length(self):
+        return self._share_file.get_length()
```
Comment (exarkun) [inline src/allmydata/storage/immutable.py:553]: and here

File: src/allmydata/storage/mutable.py [hunk 5]
```diff
@@ -412,11 +412,11 @@ def readv(self, readv):
                 datav.append(self._read_share_data(f, offset, length))
         return datav
 
-#    def remote_get_length(self):
-#        f = open(self.home, 'rb')
-#        data_length = self._read_data_length(f)
-#        f.close()
-#        return data_length
+    def get_length(self):
+        f = open(self.home, 'rb')
```
Comment (exarkun) [inline src/allmydata/storage/mutable.py:416]: and here

File: src/allmydata/storage/server.py [hunk 6]
```diff
@@ -794,6 +794,21 @@ def advise_corrupt_share(self, share_type, storage_index, shnum,
 
         return None
 
+    def get_immutable_share_length(self, storage_index: bytes, share_number: int) -> int:
+        """Returns the length (in bytes) of an immutable."""
+        si_dir = storage_index_to_dir(storage_index)
+        path = os.path.join(self.sharedir, si_dir, str(share_number))
+        bucket = BucketReader(self, path, storage_index, share_number)
```
Comment (exarkun) [inline src/allmydata/storage/server.py:801]: This could go directly to `ShareFile` I think?

**Review by exarkun — APPROVED** [review]
Thanks.  Looks good.  Some minor comments inline.  Please address those and then merge.

File: src/allmydata/storage/immutable.py (hunk 4 above)
Comment (itamarst) [reply src/allmydata/storage/immutable.py:553]: Ended up deleting this method, per later comment.

**Pull request merged**